[Maps] Show labels after saving edits while staying on Vector tiles (…

jmalmo opened the pull request:
Closes #240727 
---

## Summary
When using **Maps → Create index (Draw shapes)**, the new layer defaults to **Scaling → Use vector tiles**.  
Adding label styling (Fixed or By value) previously did **not** display labels until the user manually changed Scaling to **Limit results** and back.  
This PR makes labels appear immediately (after styling or save) on vector-tiles layers by:

- Requesting MVT tiles with **`with_labels=true`** when label styling is active  
- **Refreshing** tile sources when label styling changes and after saving edits  
- Properly routing **GEOJSON_VECTOR** layers with `scalingType: MVT` through the MVT pipeline

Elastic’s `_mvt` API returns suggested label points when `with_labels=true`; this wires Kibana to use that path and refresh tiles accordingly.  
Kibana’s **Vector layer** docs describe the scaling modes this aligns with.

> 🧠 Investigation and solution developed with the help of the large language model **Claude Sonnet 4.5**, used to analyze the root cause and assist in implementation.

---

### What changed (files)
1. **`x-pack/platform/plugins/shared/maps/public/classes/layers/vector_layer/mvt_vector_layer/mvt_vector_layer.tsx`**  
   • Snapshots the current label descriptor and `hasLabels` flag into the source request meta; any change invalidates the cached tiles and re-fetches through the MVT layer.

2. **`x-pack/platform/plugins/shared/maps/public/classes/layers/vector_layer/mvt_vector_layer/mvt_source_data.ts`**  
   • Respect `isForceRefresh` (was checking only `forceRefreshDueToDrawing`).  
   • Generate a new token only when `hasLabels` or buffer change (prevents loops).  
   • Include **`with_labels=true`** in tile requests when label styling is active (Elasticsearch `_mvt` label points).  

3. **`x-pack/platform/plugins/shared/maps/public/classes/layers/wizards/new_vector_layer_wizard/wizard.tsx`**  
   • When Create index uses MVT scaling, emit an **MvtVectorLayer** descriptor so the layer starts on the MVT pipeline and picks up the label-refresh logic.

> **Note:** `geojson_vector_layer.tsx` is restored to match `main` (no MVT source added by GeoJSON). Vector-tile rendering remains owned by `MvtVectorLayer`.

Per Elastic’s docs, the vector-tile search API can return suggested label points when `with_labels=true`; the Vector layer docs describe when to use **Use vector tiles** vs **Limit results**. This change wires Kibana’s MVT path to request tiles with labels and refresh them when styling/saving, so labels render without a scaling toggle.



---

### Why
During the Create index workflow, vector-tiles scaling is the default.  
Tiles are built server-side from indexed data, so unsaved edits and label changes don’t show until tiles are refreshed.  
By requesting tiles with labels and refreshing them on style or save, labels now render as expected without a manual scaling toggle.

---

### Before / After
| Before | After |
|:--|:--|
| Create index → Draw point → Add label → Label missing until Scaling changed | Create index → Draw point → Add label → Label appears instantly (no toggle needed) |

**Before (GIF):**  
![kibana_gif_not_working](https://github.com/user-attachments/assets/6295581d-a902-47a3-aeea-4f8dfb616793)

**After (GIF):**  
![kibana_working_gif](https://github.com/user-attachments/assets/e859c649-6eb1-4139-adea-6bd7ed5f7986)

---

### Testing
This change was verified manually; **no new unit or functional tests were added** in this PR.

**Manual validation steps**
- Add layer → Create index
- Draw a point or other shape
- Add label styling (Fixed value and By value) → labels appear immediately without toggling scaling
- Modify label text → label updates in real time
- Toggle labels on/off → tiles refresh correctly
- Save edits → labels persist after saving
- No crashes, flickering, or infinite refresh loops

**Test coverage note**
Existing Jest tests for Maps layers and source syncing still pass.  
Potential follow-ups:
- Label style change triggers `isForceRefresh=true`
- `with_labels=true` included in MVT tile requests when labels are active
- GeoJSON→MVT rendering path switching works without regression

---

## Release note
**Fix:** Labels in the **Create index** flow now render with the default **Use vector tiles** scaling as soon as label styling is applied (or after save), without requiring a scaling toggle.

---

### Checklist
- [x] No new UI strings (i18n N/A); follows EUI sentence case  
- [x] No HTTP API or security changes  
- [ ] Unit tests updated/added *(none in this PR; validated manually)*  
- [x] Include release note  
- [ ] Labels (`release_note:*`, `backport:*`, `review`, `community`) will be applied by maintainers during triage

---

### Identify risks
- **Scope:** Limited to Create index / MVT-scaling render path  
- **Performance:** `with_labels` adds small label points per tile; only used when labels are enabled  
- **Mitigation:** Refresh is scoped to the affected layer; token generation avoids loops

---

### Validation
```bash
nvm use 22.17.1
yarn kbn bootstrap
yarn lint
yarn test:jest x-pack/platform/plugins/shared/maps
yarn es snapshot --license trial
yarn start


<!--ONMERGE {"backportTargets":["8.19","9.1","9.2"]} ONMERGE-->

jmalmo committed:
[Maps] Show labels after saving edits while staying on Vector tiles (MVT)

When a Maps layer uses "Use vector tiles" scaling, labels now appear after
the user saves edits or modifies label styling, without requiring the user
to toggle scaling away from Vector tiles.

## Problem

Users adding label styling (Fixed value or By value) to ES Search layers
with "Use vector tiles" scaling would not see labels appear until they
manually toggled the scaling setting away from Vector tiles and back.

## Root Cause

Three interconnected issues prevented labels from appearing:

1. **Label change detection failure**: The style change handler checked for
   MVT_VECTOR layer type, but ES Search layers using "Use vector tiles" are
   GEOJSON_VECTOR type with scalingType=MVT property. This meant label
   changes were never detected for the most common use case.

2. **Missing force refresh support**: The MVT tile sync logic only checked
   forceRefreshDueToDrawing but ignored isForceRefresh flag used for style
   changes. Even when label changes were detected elsewhere, tiles wouldn't
   refresh.

3. **Wrong rendering path**: GEOJSON_VECTOR layers with scalingType=MVT were
   using GeoJSON rendering (fetching features as GeoJSON) instead of MVT
   rendering (requesting vector tiles). This meant the layer never requested
   tiles with with_labels=true, so Elasticsearch never generated label points.

## Solution

**1. Detect label changes for MVT-scaled layers (layer_actions.ts)**
   - Check sourceDescriptor.scalingType === SCALING_TYPES.MVT instead of
     layer type
   - Compare previous/next label configurations using JSON.stringify
   - Dispatch syncDataForLayerId with isForceRefresh=true when labels change
   - Works for both GEOJSON_VECTOR and MVT_VECTOR layer types

**2. Respect force refresh flag (mvt_source_data.ts)**
   - Added !syncContext.isForceRefresh to canSkip condition
   - Optimized refresh token generation: only create new token when hasLabels
     or buffer actually change (prevents infinite refresh loops)
   - Ensures tiles are re-fetched when label styling changes

**3. Make GeoJsonVectorLayer MVT-aware (geojson_vector_layer.tsx)**
   - _isTiled(): Returns true when source.isMvt() is true (enables vector
     tile rendering)
   - syncLayerWithMB(): Uses MapLibre vector tile source instead of GeoJSON
     source when using MVT scaling
   - _syncData(): Calls syncMvtSourceData() with hasLabels parameter for MVT
     scaling, ensuring tiles are requested with with_labels=true
   - getLayerIcon(), getBounds(), getFeatureById(),
     getStyleMetaDescriptorFromLocalFeatures(): Handle MVT mode gracefully to
     prevent crashes from accessing non-existent feature collection
   - _requiresPrevSourceCleanup(): Detect when MVT tile URL changes for
     proper layer re-rendering

## Testing

Tested with ES Search source using "Use vector tiles" scaling:
- ✅ Add Fixed value label styling → Labels appear immediately
- ✅ Add By value label styling → Labels appear immediately
- ✅ Toggle labels on/off → Tiles refresh correctly with/without labels
- ✅ Change label text while typing → Updates in real-time
- ✅ No crashes when creating layers
- ✅ No infinite refresh loops or flickering
- ✅ Layer icons and bounds work correctly
- ✅ Performance: tiles refresh only when hasLabels or buffer actually change

## Files Changed

- x-pack/platform/plugins/shared/maps/public/actions/layer_actions.ts
- x-pack/platform/plugins/shared/maps/public/classes/layers/vector_layer/geojson_vector_layer/geojson_vector_layer.tsx
- x-pack/platform/plugins/shared/maps/public/classes/layers/vector_layer/mvt_vector_layer/mvt_source_data.ts

jmalmo committed:
Fix Create index vector tile labels after styling changes

trigger a forced sync when label styling toggles on MVT-scaled layers
include label/buffer state in MVT tile refresh tokens and request with_labels=true
instantiate MvtVectorLayer descriptors from the Create index wizard when scaling stays MVT

jmalmo committed:
Move MVT label refresh trigger into layer implementation

- track the current label descriptor/hasLabels in MvtVectorLayer.syncData so tile cache invalidates on label changes
- continue honoring hasLabels/buffer in syncMvtSourceData, requesting tiles with with_labels=true only when needed
- make the Create index wizard instantiate MvtVectorLayer descriptors when scaling is MVT, so the refresh logic always applies

Testing: manual (Create index → draw feature → toggle labels)

nreese reviewed (commented):
The issue is isolated to x-pack/platform/plugins/shared/maps/public/classes/layers/wizards/new_vector_layer_wizard/wizard.tsx. Once the wizard creates the correct layer, then issue is resolved. Please revert changes to mvt_source_data and mvt_vector_layer as they are not needed.

jmalmo committed:
Address review: Make index wizard use default layer descriptor

- reuse createDefaultLayerDescriptor so MVT scaling yields an MvtVectorLayer
- drop the earlier MVT layer/source tweaks that aren’t needed once the wizard creates the correct layer

jmalmo committed:
Address review: we now call createDefaultLayerDescriptor directly with the raw descriptor fields (no intermediate ESSearchSource.createDescriptor)

jmalmo commented on x-pack/platform/plugins/shared/maps/public/classes/layers/wizards/new_vector_layer_wizard/wizard.tsx:
```diff
@@ -128,8 +128,8 @@ export class NewVectorLayerEditor extends Component<RenderWizardArguments, State
     });
-    const layerDescriptor = GeoJsonVectorLayer.createDescriptor(
-      { sourceDescriptor },
+    const layerDescriptor = createDefaultLayerDescriptor(
```
👍

nreese reviewed (approved):
changes LGTM - thanks for contributing to kibana

elasticmachine committed:
Merge branch 'main' into maps/fix-vector-tile-labels-after-save

elasticmachine commented:
Pinging @elastic/kibana-presentation (Team:Presentation)

nreese commented:
/ci

elasticmachine commented:
## :hourglass_flowing_sand: Build in-progress, with failures
* [Buildkite Build](https://buildkite.com/elastic/kibana-pull-request/builds/356362)
* Commit: b74b45cbfc6950c87125c99d0ed169cbfac79699

### Failed CI Steps
* [Check Types](https://buildkite.com/elastic/kibana-pull-request/builds/356362#019a30e5-4a66-47af-9d57-00c856335b05)

<!--buildkite-pr-comment-kibana-pull-request
{"builds":[{"buildStatus":{"state":"failing","success":false,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/356362","number":356362,"commit":"b74b45cbfc6950c87125c99d0ed169cbfac79699"}],"number":356362}
buildkite-pr-comment-->

kibanamachine committed:
Changes from node scripts/eslint_all_files --no-cache --fix

nreese commented:
/ci

nreese commented:
@elasticmachine run docs-build

kibanamachine commented:
Starting backport for target branches: 8.19, 9.1, 9.2

https://github.com/elastic/kibana/actions/runs/18928624698

kibanamachine commented:
## 💔 Some backports could not be created

| Status | Branch | Result |
|:------:|:------:|:------|
|❌|8.19|Backport failed because of merge conflicts|
|❌|9.1|Backport failed because of merge conflicts|
|✅|9.2|[<img src="https://img.shields.io/github/pulls/detail/state/elastic/kibana/241219">](https://github.com/elastic/kibana/pull/241219)|

Note: Successful backport PRs will be merged automatically after passing CI.

### Manual backport
To create the backport manually run:
```
node scripts/backport --pr 240728
```

### Questions ?
Please refer to the [Backport tool documentation](https://github.com/sorenlouv/backport)

<!--- Backport version: 9.6.6 -->
